security/cmd/node_agent/management/server.go

```diff
@@ -0,0 +1,208 @@
+}
+
+// Stop terminates the UDS.
+func (s *Server) Stop() {
+	s.done <- true
+}
+
+// WaitDone waits for a response back from Workloadhandler
+func (s *Server) WaitDone() {
+	<-s.done
+}
+
+// Serve opens the UDS channel and starts to serve NodeAgentService on that uds channel.
+func (s *Server) Serve(path string) {
+	grpcServer := grpc.NewServer()
+	pb.RegisterNodeAgentServiceServer(grpcServer, s)
+
+	var lis net.Listener
+	var err error
+	_, e := os.Stat(path)
+	if e == nil {
+		e := os.RemoveAll(path)
+		if e != nil {
+			log.Errorf("failed to %s with error %v", path, err)
+		}
+	}
+	lis, err = net.Listen("unix", path)
+	if err != nil {
+		log.Errorf("failed to %v", err)
+	}
+
+	go func(ln net.Listener, s *Server) {
+		<-s.done
+		_ = ln.Close()
+		s.CloseAllWlds()
+	}(lis, s)
+
+	_ = grpcServer.Serve(lis)
+}
+
+// WorkloadAdded defines the server side action when a workload is added.
+func (s *Server) WorkloadAdded(ctx context.Context, request *pb.WorkloadInfo) (*pb.NodeAgentMgmtResponse, error) {
+	uid := request.Attrs.Uid
+	sa := request.Attrs.Serviceaccount
+	log.Infof("workload uid = %v, service account = %v added", uid, sa)
+	if _, ok := s.handlerMap[uid]; ok {
+		status := &rpc.Status{Code: int32(rpc.ALREADY_EXISTS), Message: "Already exists"}
+		log.Infof("workload %v already exists", request)
+		return &pb.NodeAgentMgmtResponse{Status: status}, nil
+	}
+
+	// Sends CSR request to CA.
+	// TODO(inclfy): extract the SPIFFE formatting out into somewhere else.
+	id := fmt.Sprintf("spiffe://cluster.local/ns/%s/sa/%s", request.Attrs.Namespace, sa)
+	priv, csrReq, err := s.createRequest(id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create csr request for service %v %v", sa, err)
+	}
+	resp, err := s.cAClient.SendCSR(csrReq, s.pc, s.config.IstioCAAddress)
+	if err != nil {
+		return nil, fmt.Errorf("csr request failed for service %v %v", sa, err)
+	}
+
+	if err := s.ss.SetServiceIdentityPrivateKey(priv); err != nil {
+		return nil, fmt.Errorf("failed in set private key for service account %v, uid %v err %v", sa, uid, err)
+	}
+	if err := s.ss.SetServiceIdentityCert(resp.SignedCert); err != nil {
+		return nil, fmt.Errorf("failed in set cert for service account %v, uid %v err %v", sa, uid, err)
+	}
+
+	s.handlerMap[uid] = handler.NewHandler(request, s.config.WorkloadOpts)
+	go s.handlerMap[uid].Serve()
+
+	status := &rpc.Status{Code: int32(rpc.OK), Message: "OK"}
+	return &pb.NodeAgentMgmtResponse{Status: status}, nil
+}
+
+// WorkloadDeleted defines the server side action when a workload is deleted.
+func (s *Server) WorkloadDeleted(ctx context.Context, request *pb.WorkloadInfo) (*pb.NodeAgentMgmtResponse, error) {
+	uid := request.Attrs.Uid
+	if _, ok := s.handlerMap[uid]; !ok {
+		status := &rpc.Status{Code: int32(rpc.NOT_FOUND), Message: "Not found"}
+		return &pb.NodeAgentMgmtResponse{Status: status}, nil
+	}
+
+	log.Infof("Uid %s: Stop.", uid)
+	s.handlerMap[uid].Stop()
+	s.handlerMap[uid].WaitDone()
```

Comment:
Is it possible the thread is stuck on WaitDone()?

Reply:
OK. That means WaitDone() will only block the single request thread. Thanks for the investigation.